src/core/Oscillator.cpp

```diff
@@ -88,6 +106,238 @@ void Oscillator::update( sampleFrame * _ab, const fpp_t _frames,
+	// initialize s_specBuf content to zero, since the values are used in a condition inside generateFromFFT()
+	for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH * 2 + 1; i++)
+	{
+		s_specBuf[i][0] = 0.0f;
+		s_specBuf[i][1] = 0.0f;
+	}
+}
+
+void Oscillator::destroyFFTPlans()
+{
+	fftwf_destroy_plan(s_fftPlan);
+	fftwf_destroy_plan(s_ifftPlan);
+	fftwf_free(s_specBuf);
+}
+
+void Oscillator::generateWaveTables()
+{
+	//generate sine tables
+	auto sinGen = []()
+	{
+		for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
+		{
+			generateSineWaveTable(s_waveTables[WaveShapes::SineWave][i]);
+		}
+	};
+
+	//generate saw tables
+	auto sawGen = []()
+	{
+		for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
+		{
+			generateSawWaveTable(OscillatorConstants::MAX_FREQ / freqFromWaveTableBand(i), s_waveTables[WaveShapes::SawWave][i]);
+		}
+	};
+	//generate square tables
+	auto squareGen = []()
+	{
+		for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
+		{
+			generateSquareWaveTable(OscillatorConstants::MAX_FREQ / freqFromWaveTableBand(i), s_waveTables[WaveShapes::SquareWave][i]);
+		}
+	};
+	//generate triangle tables
+	auto triangleGen = []()
+	{
+		for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
+		{
+			generateTriangleWaveTable(OscillatorConstants::MAX_FREQ / freqFromWaveTableBand(i), s_waveTables[WaveShapes::TriangleWave][i]);
+		}
+	};
+	auto fftGen = []()
+	{
+		//generate moogSaw tables
+		//generate signal buffer
+		for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
+		{
+			for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; ++i)
+			{
+				Oscillator::s_sampleBuffer[i] = moogSawSample((float)i / (float)OscillatorConstants::WAVETABLE_LENGTH);
+			}
+			fftwf_execute(s_fftPlan);
+			generateFromFFT(OscillatorConstants::MAX_FREQ / freqFromWaveTableBand(i), 0.2f, s_waveTables[WaveShapes::MoogSawWave][i]);
+		}
+
+		//generate Exp tables
+		//generate signal buffer
+		for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
+		{
+			for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; ++i)
+			{
+				s_sampleBuffer[i] = expSample((float)i / (float)OscillatorConstants::WAVETABLE_LENGTH);
+			}
+			fftwf_execute(s_fftPlan);
+			generateFromFFT(OscillatorConstants::MAX_FREQ / freqFromWaveTableBand(i), 0.001f, s_waveTables[WaveShapes::ExponentialWave][i]);
+		}
+	};
+
+#if !defined(__MINGW32__) && !defined(__MINGW64__)
+	std::thread sinThread(sinGen);
+	std::thread sawThread(sawGen);
+	std::thread squareThread(squareGen);
+	std::thread triangleThread(triangleGen);
+	std::thread fftThread(fftGen);
+	sinThread.join();
+	sawThread.join();
+	squareThread.join();
+	triangleThread.join();
+	fftThread.join();
+#else
+	sinGen();
+	sawGen();
+	squareGen();
+	triangleGen();
+	fftGen();
+#endif
```

Comment:
If you really need threading, you can use the `mingw-std-threads` submodule under `src/3rdparty` for MinGW. See `plugins/vst_base/RemoteVstPlugin/CMakeLists.txt` or https://github.com/LMMS/lmms/pull/4443/files#diff-c44352fef68563cb31b7e5fcb38a4739c8fd8c60017bca82d538e2ca120be2da for details.

Reply:
The threading is not really _needed_, but the speedup is quite noticeable so I think it would be good to keep it. I did not want to over-complicate things to support a broken compiler, but if including the `3rdparty/` `mingw-std-threads` works and does not break anything, that would be nice. I will try to add it, thanks for the tip.

Reply:
I tried to use the code from PR 4443 and it did not work, as if the include did not happen:
`/root/project/src/core/Oscillator.cpp:323:7: error: 'thread' is not a member of 'std'`

So I either missed something or .. I don't know. I don't really understand much of what's happening in `cmake`, so this is quickly evolving exactly into what I've been trying to avoid (wasting time on trying to fix a broken compiler). I will keep the original "ignore mingw" version in place and perhaps try to enable it again after 4443 is merged with all the required `cmake` code.

Reply:
> but the speedup is quite noticeable

How does the wavetable generation take without threading?

Reply:
IIRC, one of the waveforms is especially slow and takes around 2 seconds (on a few years old i7-5500U laptop), another one or two took maybe around 1 second each, and the FFT generated waveforms were very fast, just a fraction of a second.
So essentially it is something like 4 seconds without multithreading and 2 seconds with parallel initialization (since the slowest wave is a bottleneck). For comparison, without this PR LMMS only takes around 2 s to start, so it is a relatively noticeable delay either way.

Reply:
I'm OK with threads. Though I hope that if we change the for-loops from O(n^3) to O(n^2) (see other comments), the performance will increase.

Reply:
After optimizations, the initialization takes ~400 ms in series, out of which the saw wave takes around 180 ms (measured on i3-2100, 2 cores / 4 threads). So on a 4 core machine it should save around 50 % of time (since the saw wave will be the bottle-neck). That should be still useful on computers like Raspberry Pi: version 3B seems to be at least 10 times slower than i3-2100, so it could again end up being a difference between 2 and 4 seconds.

So keeping the threads and marking resolved.